File: J9Ch/src/S_5_Dynamic_Problem_I/Optional_7/_435_Post_Office_Problem_Hard.py

class Solution:
    # @param {int[]} A an integer array
    # @param {int} k an integer
    # @return {int} an integer
    def get_w(self):
        for i in xrange(1, self.n - 1):
            for j in xrange(i, self.n):
                if (i+j)%2 == 1:
                    self.w[i][j] = self.s[j]-self.s[(i+j)/2]*2+self.s[i-1]
                else:
                    self.w[i][j] = self.s[j]-self.s[(i+j)/2]-self.s[(i+j)/2-1]+self.s[i-1]

    def get_dp(self):
        dp = [[0 for j in xrange(self.n + 10)] for i in xrange(self.k  + 10)]
        p = [[0 for j in xrange(self.n + 10)] for i in xrange(self.k  + 10)]
        for i in xrange(1, self.n):
            dp[1][i] = self.w[1][i]
            p[1][i] = 0

        n = self.n - 1
        for c in xrange(2, self.k + 1):
            p[c][n+1] = n
            for i in xrange(n, 0, -1):
                tmp=0x7fffffff
                k = 0
                for j in xrange(p[c-1][i], p[c][i+1] + 1):
                    if dp[c-1][j] + self.w[j+1][i] < tmp:
                        tmp = dp[c-1][j] + self.w[j+1][i]
                        k = j
                dp[c][i]=tmp
                p[c][i]=k

        return dp[self.k][self.n - 1]

    def postOffice(self, A, k):
        # Write your code here
        A = sorted(A)
        if k >= len(A):
            return 0

        self.n = len(A) + 1
        self.k = k
        self.s = [0]
        self.w = [[0 for j in xrange(self.n + 10)] for i in xrange(self.n  + 10)]
        for x in A:
            self.s.append(self.s[-1] + x)

        self.get_w()
        return self.get_dp()

File: J9Ch/src/S_5_Dynamic_Problem_I/Optional_7/_435_Post_Office_Problem_Hard.cpp

int f[1500][1500], w[1500][1500], a[1500], K[1500][1500];
class Solution {
public:
    /**
     * @param A an integer array
     * @param k an integer
     * @return an integer
     */
    int N, P;
	vector<int> tmp;
    void init() {
        int n = tmp.size();
        a[0] = 0;
        for (int i = 1; i <= n; i ++) {
            a[i] = a[i - 1] + tmp[i-1];
        }
    }

    int getw(int x, int y)  {
        int t = (x + y) / 2;
        return a[y] - a[t] - (y - t) * tmp[t - 1] + (t - x) * tmp[t - 1] - (a[t - 1] - a[x - 1]);
    }

    void solve(int N, int P) {
        for (int i = 0; i <= N; i ++) {
            f[i][i] = 0;
            K[i][i] = i;
        }
		int j;
        for (int p = 1; p <= N - P; p ++) {
            for (int i = 0; (j  = i + p) <= N; i ++)
                f[i][j] = 0x7fffffff / 3;
            int t;
            for (int i = 1; (j = i + p) <= N; i ++) {
                for (int k = K[i][j - 1]; k <= K[i + 1][j]; k ++)
                    if ((t = f[i - 1][k - 1] + getw(k, j)) < f[i][j]) {
                        f[i][j] = t;
                        K[i][j] = k;
                    }
            }
        }
    }

    void getAns(int k, int r, vector<int>& ret) {
        if ( k == 0 || r <= 0)
            return;
        int p = K[k][r];
        getAns(k - 1, p - 1, ret);
        ret.push_back(tmp[(p + r) / 2 - 1]);
    }

    int postOffice(vector<int>& A, int k) {
        // Write your code here
        sort(A.begin(), A.end());
		tmp = A;
        int n = A.size();
        init();
        solve(n, k);

        vector<int> result;
        getAns(k, n, result);
        return f[k][n];
    }
};